### PR TITLE
Fix loose focus for Web/Desktop

### PR DESCRIPTION
**Pull request by narefyev91**
### Details
Amount field loose focus clicking outside

### Fixed Issues
$ https://github.com/Expensify/App/issues/16511
$ https://github.com/Expensify/App/issues/16511#issuecomment-1487262552

### Tests
1. Launch expensify app
2. Login with any account
3. Tap on the FAB button -> Send money
4. Enter numbers
5. Click outside of numpad
6. Enter numbers

- [ ] Verify that no errors appear in the JS console

### Offline tests
Nothing needed

### QA Steps
Platform specific steps:
#### Web/Desktop:
1. Launch expensify app
2. Login with any account
3. Tap on the FAB button -> Send money
4. Enter numbers
5. Click outside of numpad

_Expected result - amount input will not loose focus_

#### mWeb:
1. Launch expensify app
2. Login with any account
3. Tap on the FAB button -> Send money
4. Enter numbers
5. Click outside of numpad
6. Enter numbers

_Expected results - amount input will be automatically focused - new typed number will be shown_

#### Native IOS, Android:
1. Launch expensify app
2. Login with any account
3. Tap on the FAB button -> Send money
4. Enter numbers
5. Click outside of numpad

_Expected result - amount input will not loose focus_

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x]  Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/28352309/228561609-77b46500-94d8-4f9c-8c86-88172e520be9.mp4

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/28352309/228561101-9c9ee996-6232-4bce-adbe-ac4679190848.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/28352309/228561169-01d2f96e-8a86-45d5-9edf-55c7c1481e31.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/28352309/228561198-8dfa72be-7f95-41ed-9244-88b9766995e8.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/28352309/228561226-de772e76-86cb-4bee-9a0a-47fc3a0cdc13.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/28352309/228561254-e8e50bc9-a1a5-47bb-88eb-62b6ff6d023d.mov

</details>

### CODE REVIEW

**Review by mountiny — DISMISSED** [review]
Looking good, thanks, @narefyev91 I think we could improve the testing steps to also include what to **verify** by the QA team. Could you make the changes also platform specific.

**Comment by narefyev91** [comment]
> Looking good, thanks, @narefyev91 I think we could improve the testing steps to also include what to **verify** by the QA team. Could you make the changes also platform specific.

Sure - updated description based on platforms

**Review by mountiny — DISMISSED** [review]
Looks good to me, all yours @0xmiroslav Thanks!

File: src/pages/iou/steps/IOUAmountPage.js [hunk 1]
```diff
@@ -267,6 +268,12 @@ class IOUAmountPage extends React.Component {
                 ]}
                 >
                     <TextInputWithCurrencySymbol
+                        onBlur={(event) => {
+                            const relatedTargetId = lodashGet(event, 'nativeEvent.relatedTarget.id');
+                            if (!relatedTargetId) {
+                                this.focusTextInput();
+                            }
```
[the remark this answers is not in the document]
Comment (narefyev91) [reply src/pages/iou/steps/IOUAmountPage.js]: Yup let's have the same approach and not use navigation handler to focus an input

**Comment by narefyev91** [comment]
> Just confirming this one: [#16673 (comment)](https://github.com/Expensify/App/pull/16673#discussion_r1152798599)

@0xmiroslav @mountiny resolved latest comment

**Comment by narefyev91** [comment]
> > in that case input should loose focus if we interact with header? @0xmiroslav
> 
> I am not quite sure but I think at least when interact with header title. Anyone who wants to copy & paste "Split bill" text? 😄

Ok, but not sure if it will be possible to still use onBlur handler - event from onBlur will not contain information about something which placed before text input in DOM tree.

**Comment by narefyev91** [comment]
> Yeah we should mimic the behaviour in production

@0xmiroslav @mountiny came to one more solution which mostly covering all possible cases and not make any flickering regressions as well, and make header selectable
```
                <View
                    nativeID={this.amountViewID}
                    onMouseDown={(event) => {
                        const relatedTargetId = lodashGet(event, 'nativeEvent.srcElement.id');
                        if (_.contains([this.amountViewID], relatedTargetId)) {
                            event.preventDefault();
                        }
                    }}
```
We add new nativeID, and indicate if some clicks exactly in the dark area, if yes and nativeID is equal to which section we clicked - we will not remove focus. Clicking on currency selector and button will not make any UI visual issues because both have nativeID and they will not be pass inside if condition.

**Comment by narefyev91** [comment]
> @narefyev91 doesn't that approach work on `onBlur`? So focus only when meet `if (_.contains([this.amountViewID], relatedTargetId)) {` condition

nope because onBlur event know nothing about upper View

**Comment by narefyev91** [comment]
> > nope because onBlur event know nothing about upper View
> 
> correct, it didn't work. And I thought it also won't work on `onMouseDown`. you tested?

yup tested

https://user-images.githubusercontent.com/28352309/228875320-39726e04-5266-4e23-818b-d9b638952209.mp4

**Comment by narefyev91** [comment]
> @narefyev91 I think we already concerned about this:
> 
> * let's say "Split bill" text is selected
> * click black area
> * preventDefault() is applied and text still keeps selected instead of focusing input

```
                    onMouseDown={(event) => {
                        const relatedTargetId = lodashGet(event, 'nativeEvent.srcElement.id');
                        if (_.contains([this.amountViewID], relatedTargetId)) {
                            event.preventDefault();
                            if (!this.textInput.isFocused()) {
                                this.textInput.focus();
                            }
                        }
                    }}
```

https://user-images.githubusercontent.com/28352309/228877190-fa8dd712-3e1a-48f0-b979-d5844d41ed97.mp4

**Comment by mountiny** [comment]
niiiice

**Comment by narefyev91** [comment]
> Is this behavior fine?
@0xmiroslav 
lol i think you are trying something really strange behaviours
When you do a select numbers, and after select something on header, go back and click on dark area - your input is still focused, and if you will try to typing if will work

**Comment by mountiny** [comment]
When you clicked next, it did not go to next page?

**Comment by narefyev91** [comment]
> When you clicked next, it did not go to next page?

he is not clicking next, he click on next and hold the mouse button, and move cursor to dark area and take out finger from mouse lol

**Comment by narefyev91** [comment]
> Keeps selection range even when go to another page (currency selection) and return back. But when click Next button and return back, selection is cleared. So inconsistency issue

Yup this is same behaviour that already exist in main branch - should we also fix that?

**Comment by narefyev91** [comment]
> @narefyev91 let's remove unused nativeIDs

@0xmiroslav done

File: src/pages/iou/steps/IOUAmountPage.js [hunk 2]
```diff
@@ -180,7 +181,7 @@ class IOUAmountPage extends React.Component {
      * @param {String} key
      */
     updateAmountNumberPad(key) {
-        this.focusTextInput();
+        this.textInput.focus();
```
[the remark this answers is not in the document]
Comment (narefyev91) [reply src/pages/iou/steps/IOUAmountPage.js]: @0xmiroslav @mountiny maybe that's only on my devices - but as i see caret not shown everywhere when we have autofocus on main branch (our changes is not influence and done correctly)

https://user-images.githubusercontent.com/28352309/229049917-f62e9cc1-ec86-4206-b7b3-678887664466.mp4

**Comment by narefyev91** [comment]
> @narefyev91 losing focus when click numpad gaps

add fix for it

https://user-images.githubusercontent.com/28352309/229062723-7f7c4a08-5e11-4cea-b60d-871bccdb2893.mov

**Comment by narefyev91** [comment]
> > @narefyev91 losing focus when click numpad gaps
> 
> @0xmiroslav @mountiny  It was initially request from Vit to focus input when clicking on numpad item (if we loose focus between buttons), in any case add fix for it

**Comment by narefyev91** [comment]
> I think [last solution](https://github.com/Expensify/App/pull/16673/commits/2816e325acc0986d36e2ed9ed298f5f0f156d48c) made it complex. Is it just to fix `losing focus when click numpad gaps`? Any bad effect if we move this to wrapper? wrapper

we can move this only to this wrapper
`                <View onMouseDown={e => e.preventDefault()} style={[styles.w100, styles.justifyContentEnd, styles.pageWrapper]}>
                    {DeviceCapabilities.canUseTouchScreen()
                        ? (
                            <BigNumberPad
                                numberPressed={this.updateAmountNumberPad}
                                longPressHandlerStateChanged={this.updateLongPressHandlerState}
                            />
                        ) : <View />}

                    <Button
                        success
                        style={[styles.w100, styles.mt5]}
                        onPress={() => this.props.onStepComplete(this.state.amount)}
                        pressOnEnter
                        isDisabled={!this.state.amount.length || parseFloat(this.state.amount) < 0.01}
                        text={this.props.translate('common.next')}
                    />
                </View>`
Not really sure why it's started to be complex, we have one function which will cover any case now, and will cover if we add any new component to this page, if anything on the page will be changed we will need to add e.preventDefault rather that controlling based on nativeIDs

**Comment by narefyev91** [comment]
> > if anything on the page will be changed we will need to add e.preventDefault
> 
> It's same we will need to add nativeID. So no perfect general solution
> 
> > It was initially request from Vit to focus input when clicking on numpad item (if we loose focus between buttons)
> 
> So I agree with this and leave gap case as is for now. So I think just reverting last commit should be fine.

but why do we need to rollback commit if a new one covers also possibility not loose focus between numbers and make flow working identical ?

File: src/pages/iou/steps/IOUAmountPage.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (narefyev91) [reply src/pages/iou/steps/IOUAmountPage.js]: https://github.com/Expensify/App/issues/10414 - this is initial issue for loose focus

**Comment by narefyev91** [comment]
> > but why do we need to rollback commit if a new one covers also possibility not loose focus between numbers and make flow working identical ?
> 
> I know that works, but I feel like it's overkill. Any other part in our codebase which added multiple nativeIDs to handle such case?

Screenshot 2023-03-31 at 14 30 34

**Comment by narefyev91** [comment]
@0xmiroslav @mountiny ok what would be the last step - i rollback last commit and we good to go?

**Comment by mountiny** [comment]
> ok what would be the last step - i rollback last commit and we good to go?

Would that mean clicking between the numpads on mobile will leave focus? If yes we should not do that, thats a major UX bad

**Comment by narefyev91** [comment]
> > ok what would be the last step - i rollback last commit and we good to go?
> 
> Would that mean clicking between the numpads on mobile will leave focus? If yes we should not do that, thats a major UX bad

that means - that on mobile web - when user click between button - input will loose focus (if i revert that change)

**Comment by mountiny** [comment]
Yes we dont want that, its easy to fat finger between the numbers and it will be bad UX if user will have to click back to the number

**Review by 0xmiroslav — APPROVED** [review]

**Review by mountiny — APPROVED** [review]
Thank you both for getting this over the finish line

**Comment by abdulrahuman5196** [comment]
Unable to Remove Request/Send Money Instantly by Holding the Remove Button

his issue was caused by a series of changes of IOU Money Keypad files. 
https://github.com/Expensify/App/pull/15710
https://github.com/Expensify/App/pull/16394
https://github.com/Expensify/App/pull/16673

**Pull request merged**